### PR TITLE
A0-3509: Purge chain-extension: frontend

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
# Description

Do you remember our shady and dirty chain extension? Well, we are back.

Since OBCE is no longer maintained, we have to rework the extension from the very ground. To avoid huge diff, I divided the work into couple of separate changes. This PR:
 - removes all OBCE references
 - brings back plain frontend part (i.e. the API for the contracts)
 - creates an example contract for testing purposes

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- I have added tests
- I have made neccessary updates to the Infrastructure
- I have made corresponding changes to the existing documentation
- I have created new documentation

### CODE REVIEW

File: baby-liminal-extension/src/error_codes.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// ---- `verify` errors ----------------------------------------------------------------------------
+const BABY_LIMINAL_VERIFY_BASE: u32 = 11_000;
+pub const BABY_LIMINAL_VERIFY_SUCCESS: u32 = BABY_LIMINAL_VERIFY_BASE;
+pub const BABY_LIMINAL_VERIFY_DESERIALIZING_PROOF_FAIL: u32 = BABY_LIMINAL_VERIFY_BASE + 1;
+pub const BABY_LIMINAL_VERIFY_DESERIALIZING_INPUT_FAIL: u32 = BABY_LIMINAL_VERIFY_BASE + 2;
+pub const BABY_LIMINAL_VERIFY_UNKNOWN_IDENTIFIER: u32 = BABY_LIMINAL_VERIFY_BASE + 3;
+pub const BABY_LIMINAL_VERIFY_DESERIALIZING_KEY_FAIL: u32 = BABY_LIMINAL_VERIFY_BASE + 4;
+pub const BABY_LIMINAL_VERIFY_VERIFICATION_FAIL: u32 = BABY_LIMINAL_VERIFY_BASE + 5;
+pub const BABY_LIMINAL_VERIFY_INCORRECT_PROOF: u32 = BABY_LIMINAL_VERIFY_BASE + 6;
+pub const BABY_LIMINAL_VERIFY_ERROR_UNKNOWN: u32 = BABY_LIMINAL_VERIFY_BASE + 7;
```
Comment (kostekIV) [inline baby-liminal-extension/src/error_codes.rs:25]: nit: `BABY_LIMINAL` prefix is kinda redundant and create a lot of noise.

File: baby-liminal-extension/src/error_codes.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+//! Error codes for the baby-liminal-extension pallet.
```
Comment (kostekIV) [inline baby-liminal-extension/src/error_codes.rs:1]: nit: arent those a status code? Since success variant is amoung them

**Review by kostekIV — APPROVED** [review]

File: baby-liminal-extension/src/error_codes.rs (hunk 1 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/src/error_codes.rs]: sure, removed

File: baby-liminal-extension/src/error_codes.rs (hunk 2 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/src/error_codes.rs]: renamed, thanks

File: baby-liminal-extension/src/api.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    /// Directly call `pallet_baby_liminal::store_key`.
+    #[ink(extension = 41)]
+    fn store_key(
+        origin: AccountId,
+        identifier: VerificationKeyIdentifier,
+        key: Vec<u8>,
+    ) -> Result<(), BabyLiminalError>;
+
+    /// Directly call `pallet_baby_liminal::verify`.
+    #[ink(extension = 42)]
```
Comment (deuszx) [inline baby-liminal-extension/src/api.rs]: Can we use `extension_ids` here? Or does the macro won't work with variables?

File: baby-liminal-extension/test_contract/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+version = "0.1.0"
+authors = ["Cardinal"]
+edition = "2021"
+homepage = "https://alephzero.org"
+repository = "https://github.com/aleph-zero-foundation/aleph-node"
+license = "Apache-2.0"
+publish = false
+
+[dependencies]
+ink = { version = "4.2.0", default-features = false }
```
Comment (deuszx) [inline baby-liminal-extension/test_contract/Cargo.toml:12]: Why not `=4.3.0`?

File: baby-liminal-extension/src/api.rs (hunk 3 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/src/api.rs]: please take a look at the current code version (I added note in the latest commit)

File: baby-liminal-extension/test_contract/lib.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        }
+
+        #[ink::test]
+        fn it_works() {
+            register_chain_extension(MockedStoreKeyExtension);
+
+            let test_contract = TestContract::new();
+            test_contract.call_ce();
+        }
```
Comment (deuszx) [inline baby-liminal-extension/test_contract/lib.rs]: 🤯 

This doesn't actually test the chain extension though, right?

File: baby-liminal-extension/src/lib.rs [hunk 6]
```diff
@@ -5,13 +5,13 @@ compile_error!(
     "Features `ink` and `runtime` are mutually exclusive and cannot be enabled together"
 );
 
-#[cfg(feature = "ink")]
-pub mod api;
 pub mod error_codes;
 pub mod extension_ids;
+#[cfg(feature = "ink")]
+pub mod frontend;
```
Comment (deuszx) [inline baby-liminal-extension/src/lib.rs:16]: Too much time spent with SNARKs I think :) I liked `api` more but I'll accept this as well.

File: baby-liminal-extension/test_contract/Cargo.toml (hunk 4 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/test_contract/Cargo.toml]: bumped

File: baby-liminal-extension/test_contract/lib.rs (hunk 5 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/test_contract/lib.rs]: no, it just tests API and delegates call to the mock in the offchain engine

File: baby-liminal-extension/src/api.rs (hunk 3 above)
Comment (deuszx) [reply baby-liminal-extension/src/api.rs]: Yeah, I just saw it https://github.com/Cardinal-Cryptography/aleph-node/pull/1498/commits/676e9ff7f2f1be8c522dbfaaa269e55cdb70cca7.

File: baby-liminal-extension/src/lib.rs (hunk 6 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/src/lib.rs:16]: yeah, but there is also the second part, which in this lore will be just `backend` (sounds better than `impl` :P)

File: baby-liminal-extension/test_contract/lib.rs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        struct MockedStoreKeyExtension;
+        impl ink::env::test::ChainExtension for MockedStoreKeyExtension {
+            fn func_id(&self) -> u32 {
+                baby_liminal_extension::extension_ids::STORE_KEY_EXT_ID
+            }
+
+            fn call(&mut self, _: &[u8], _: &mut Vec<u8>) -> u32 {
+                baby_liminal_extension::status_codes::STORE_KEY_SUCCESS
+            }
+        }
```
Comment (deuszx) [inline baby-liminal-extension/test_contract/lib.rs:40]: Interesting, so the test framework allows for registering multiple chain extensions, independently, while the "regular CE" requires to combine all of them together.

**Review by deuszx — APPROVED** [review]

File: baby-liminal-extension/test_contract/lib.rs (hunk 7 above)
Comment (pmikolajczyk41) [reply baby-liminal-extension/test_contract/lib.rs:40]: no no, it registers single function, not the whole chain extension, which can gather multiple functions behind a single trait

Comment (deuszx) [reply baby-liminal-extension/test_contract/lib.rs:40]: but you could register multiple `struct`s that are implementing ChainExtension trait and call all of them from the contract.

**Pull request merged**